### PR TITLE
Specify units for ADC HIL

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

This pull request updates the ADC HIL. See #1008 for more details.

### Testing Strategy

This pull request was tested by compiling.

### TODO or Help Wanted

n/a

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

File: kernel/src/hil/adc.rs [hunk 1]
```diff
@@ -14,19 +14,32 @@ pub trait Adc {
...
+
+    /// Function to ask the ADC how many bits of resolution are in the samples
+    /// it is returning.
+    fn get_resolution_bits(&self) -> usize;
+
+    /// Function to ask the ADC what reference voltage it used when taking the
+    /// samples. This allows the user of this interface to calculate an actual
+    /// voltage from the ADC reading.
+    ///
+    /// The returned reference voltage is in millivolts.
```
Comment (phil-levis) [inline kernel/src/hil/adc.rs]: is in millivolts, or 0 if unknown.

Comment (ppannuto) [reply kernel/src/hil/adc.rs]: Would it be more "rust"-y to return an `Option<usize>` and `None` if unknown?

Comment (brghena) [reply kernel/src/hil/adc.rs]: Either of these seems fine to me. `get_resolution_bits` should work the same way though.

Comment (brghena) [reply kernel/src/hil/adc.rs]: Hmm, why should the reference voltage be optional then? If the suggestion is that the reference could be an external pin that the ADC doesn't know, that could be passed into the driver at init time by the board `main.rs` configuration.

Comment (bradjc) [reply kernel/src/hil/adc.rs]: - That value could instead be passed to whatever would be using `get_reference_voltage()`.
- If the HW doesn't know, then conceptually it makes sense that it should return that it doesn't know.
- The reference voltage could change at runtime.

**Comment by alevy** [comment]
I changed the title of the PR to be more descriptive, but I'd also greatly appreciate a more verbose description of what this PR actually does for posterity. A reference is nice, but I actually had to look at the code to figure out what change this PR decided to actually make.

**Comment by alevy** [comment]
@phil-levis I think this is what you'd been advocating for, right? Does this seem good?

**Comment by ppannuto** [comment]
Following @alevy's lead from #1044, I've added a `CHANGELOG` entry for these updates as well. Sadly, which gets merged first will conflict the other, but it should be a quick and easy fixup.

**Comment by ppannuto** [comment]
Blocked on #1078

**Comment by phil-levis** [comment]
Almost there. But I need more than 48 hours (including a holiday) to be able to respond to something like this. This should not have been merged.

    fn sample_ready(&self, sample: u16);

This needs to have a ReturnCode parameter, in case the sample is bad in some way. This could happen, for example, if you call stop_sampling() when you are in the midst of an I2C operation to an off-board I2C. It might be much simpler for the implementation to wait for the I2C operation to complete but not issue a command to read a value, and signal an event with bad data. Otherwise, if you stop then try to sample again, the sample again might fail because the I2C operation is outstanding.

I.e., stop_sampling can return EBUSY but have sample_ready give you no data. It’s important to be able to do this so you can correctly synchronize the two state machines (caller and callee).

Phil

**Comment by ppannuto** [comment]
The addition of the CHANGELOG documenting the change came in 48 hours ago, but that's just a bit of descriptive text. All of the HIL changes have been here for 9 days with no comments but approvals. That's past the one-week "no comment = approval" period we have for significant PRs, which is why I merged.

**Comment by alevy** [comment]
> This needs to have a ReturnCode parameter, in case the sample is bad in some way. This could happen, for example, if you call stop_sampling() when you are in the midst of an I2C operation to an off-board I2C.

I think you probably mean an off-chip *__ADC__*, in which case a `ReturnCode` following the logic discussed in #1052 would probably make sense. But this HIL is not for peripheral ADCs. Is there a case where a ReturnCode would be useful in an on-chip ADC?

**Comment by phil-levis** [comment]
Should there be separate HILs for on-chip vs. off-chip ADCs? That seems not very hardware independent. I think the one place we agreed this was the right decision was GPIO, because forcing every GPIO operation to be asynchronous would be extremely onerous. 

Examples of errors that the event might want to give you:
  - a sample was missed (LOVR, 38.6.18 in SAM4L data sheet)
  - the ADC was turned off between when the sample started and acquisition started
  - the ADC sample was cancelled before acquisition

**Comment by alevy** [comment]
> Should there be separate HILs for on-chip vs. off-chip ADCs?

I don't know what ADCs look like well enough, but if there is enough of a difference in how the MCU interacts with them, then absolutely YES

> That seems not very hardware independent.

It's not. HIL stands for Hardware _Interface_ Layer (not Independence). It's a set of interfaces to very similar hardware.

> A sample was missed (LOVR, 38.6.18 in SAM4L data sheet)

OK, totally! This is an overrun basically, because the client ask for too many values or something? I don't know ADCs well enough, but if this isn't an error that can be caught earlier, it seems totally reasonably to put it in the callback.

> the ADC was turned off between when the sample started and acquisition started

Wouldn't either the client know it turned it off or the start function be able to immediately return that the ADC is off? Maybe I'm misunderstanding the scenario.

> the ADC sample was cancelled before acquisition

Same. The client should know it cancelled the acquisition.

**Comment by phil-levis** [comment]
_Hardware Interface Layer (not Independence). It's a set of interfaces to very similar hardware._

Oh! OK.

I guess the general point is that the asynchrony caused by a bus is very similar to that caused by virtualization. In purely asynchronous I/O systems, this means that an error returned in initiation can deferred to the callback.

_Wouldn't either the client know it turned it off or the start function be able to immediately return that the ADC is off? Maybe I'm misunderstanding the scenario._

What if more than one capsule has access to this trait? Or what if there is a bug in the client, such that it incorrectly turns off the ADC during a sample? Just as a read/write in POSIX file I/O will return with an error code, you should get a callback with an error code. You don't want to block forever.

**Pull request merged**